data/blog/should-you-could-you-aot.md

```diff
@@ -0,0 +1,190 @@
+
+In our case, we were certainly not good to go... we had hundreds of warnings. Luckily, many of the warnings were related so really there were only a handful of big problems we had to solve:
+
+1. Serialization
+2. Configuration Bindings
+3. Dependencies on libraries that aren't AOT compatible
+4. Miscellaneous uses of reflection
+
+## Serialization
+
+At a high level, serialization is about taking a data structure and converting it into some bytes for storage or transmission... later deserialization is about converting those bytes back into the original data structure.
+
+For years, the .NET community has relied on [Newtonsoft.Json](https://www.newtonsoft.com/json) to solve this. It's a great library that allows you to serialize and deserialize almost any data structure you can imagine. However, it relies heavily on reflection and it's not AOT compatible.
+
+When compiling AOT then you'll need to use [System.Text.Json](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/source-generation?pivots=dotnet-8-0), which relies on source code generators instead of reflection. The methods in `System.Text.Json` commonly accept a `JsonSerializerContext`, which can be used to tell the serializer how to handle specific types (and also not to trim those types).
+
+You can see an example of a [custom JsonSerializerContext that we created for the Sentry SDK here](https://github.com/getsentry/sentry-dotnet/blob/75843685494d04e0e955e2e48b40ca93d61aa8f6/src/Sentry/Internal/Extensions/JsonExtensions.cs#L888-L893):
+
+```csharp
+[JsonSerializable(typeof(GrowableArray<int>))]
+[JsonSerializable(typeof(Dictionary<string, bool>))]
+[JsonSerializable(typeof(Dictionary<string, object>))]
+internal partial class SentryJsonContext : JsonSerializerContext
+{
+}
+```
+
+This custom context ensures that when we compile the Sentry SDK to a Nuget package, the compiler knows not to trim the `GrowableArray<int>`, `Dictionary<string, bool>` or `Dictionary<string, object>` types that we know get serialized by the Sentry SDK.
+
+Not everything that is possible with Newtonsoft.Json is possible with System.Text.Json however.
+
+For example, the Sentry SDK includes something called `Contexts` that gets sent with events such as crash reports through to the Sentry backend. It's [possible for users to add their own custom context](https://github.com/getsentry/sentry-dotnet/blob/a34e9844228142bd59f4d454f669207fa9b472cc/src/Sentry/Contexts.cs#L207-L208) as a `KeyValuePair<string, object>` and we have no idea what types they might supply as custom context. As such, we had to give users a way to supply their own `SentryJsonContext` for any custom types that they might be sending with Sentry events.
```

Comment:
They do but using `Newtonsoft.Json` uses reflection for this and it won't work with trimming (since there's no equivalent of the `JsonSerializationContext` to let the compiler know which types might dynamically serialized. 

I see what you're saying though. We were previously using different overloads of System.Text.Json that used dynamic code... so maybe it's more accurate to say that "Not everything that is possible when serializing using dynamically generated code is possible with the new System.Text.Json source generators however". 

I'll push a change shortly.